cni-plugin/linkerd-cni-validation/src/main.rs

```diff
@@ -0,0 +1,170 @@
+        let listen_addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), outbound_port);
+        let listener = TcpListener::bind(listen_addr)
+            .await
+            .expect(&format!("Failed to bind server to {}", listen_addr));
+        info!(server_addr=%listen_addr, "listening for incoming connections");
+
+        if let Err(_) = ready_tx.send(Ok(())) {
+            error!(server_addr=%listen_addr, "failed to send 'ready' signal, receiver dropped");
+            return Err::<(), anyhow::Error>(anyhow!(
+                "failed to bind server: listen_addr={}",
+                listen_addr
+            ));
+        }
+
+        let handler = tokio::spawn(async move {
+            loop {
+                let (mut stream, client_addr) = listener
+                    .accept()
+                    .await
+                    .expect("failed to establish connection");
+                info!(%client_addr, "accepted connection");
+                tokio::spawn(async move {
+                    let bytes = REDIRECT_RESPONSE.as_bytes();
+                    match stream.write_all(&bytes).await {
+                        Ok(_) => debug!(%client_addr, "wrote {:?} bytes", &bytes.len()),
+                        Err(e) => error!(%client_addr, "failed to write bytes to client: {:?}", e),
+                    }
+                });
+            }
+        })
+        .instrument(tracing::info_span!("srv_handler"));
+
+        tokio::select! {
+            _shutdown = shutdown_rx.signaled() => {
+                return Ok(())
+            }
+            _v = handler => {
+                return Ok(())
+            }
+        };
+    })
+    .instrument(tracing::info_span!("validation_server"));
+
+    let client_handle = tokio::spawn(async move {
+        let timeout = std::time::Duration::from_secs(120);
+        if let Err(_) = tokio::time::timeout(timeout, ready_rx).await {
+            return Err::<(), anyhow::Error>(anyhow!(
+                "timed-out (120s) waiting for server to be ready"
+            ));
+        };
+
+        let target_addr = SocketAddr::new(IpAddr::V4(target_host), target_port);
+        let mut stream = {
+            debug!(original_dst = %target_addr, "building validation client");
+            let socket = TcpStream::connect(target_addr)
+                .await
+                .with_context(|| format!("failed to connect: {}", target_addr))?;
+            let peer = socket.peer_addr()?;
+            debug!(original_dst = %peer, "client connected to validation server");
+            socket
+        };
+
+        tokio::select! {
+            is_readable = stream.readable() => {
+                if let Err(e) = is_readable {
+                    anyhow::bail!("cannot read off client socket {}", e);
+                }
+            }
+
+            () = tokio::time::sleep(timeout) => {
+                anyhow::bail!("timed-out (120s) waiting for socket to become readable");
```

Comment:
similarly, it would be nice to use `timeout` in the error message here:
```suggestion
                anyhow::bail!("timed-out ({:?}) waiting for socket to become readable", timeout);
```